File: eel-core/src/main/scala/io/eels/component/parquet/ParquetReaderSupport.scala
package io.eels.component.parquet

import com.typesafe.config.ConfigFactory
import com.typesafe.scalalogging.slf4j.StrictLogging
import io.eels.SchemaType
import io.eels.component.hive.Predicate
import org.apache.avro.generic.GenericRecord
import org.apache.avro.{Schema, SchemaBuilder}
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
import org.apache.parquet.avro.{AvroParquetReader, AvroReadSupport}
import org.apache.parquet.filter2.compat.FilterCompat
import org.apache.parquet.filter2.compat.FilterCompat.{Filter, NoOpFilter}
import org.apache.parquet.hadoop.ParquetReader

object ParquetReaderSupport extends StrictLogging {

  val config = ConfigFactory.load()

  lazy val parallelism = {
    val parallelism = config.getInt("eel.parquet.parallelism")
    logger.debug(s"Creating parquet reader with parallelism = $parallelism")
    parallelism.toString
  }

  /**
    * Creates a new reader from the given path. If projection is set then a projected schema is generated
    * from the given schema.
    */
  def createReader(path: Path,
                   isProjection: Boolean,
                   predicate: Option[Predicate],
                   schema: io.eels.Schema): ParquetReader[GenericRecord] = {
    require(!isProjection || schema != null, "Schema cannot be null if projection is set")

    def projection: Schema = {
      val builder = SchemaBuilder.record("row").namespace("namespace")
      schema.columns.foldLeft(builder.fields) { (fields, col) =>
        val schemaType = col.`type`
        val name = col.name
        schemaType match {
          case SchemaType.BigInt => fields.optionalLong(name)
          case SchemaType.Boolean => fields.optionalBoolean(name)
          case SchemaType.Double => fields.optionalDouble(name)
          case SchemaType.Float => fields.optionalFloat(name)
          case SchemaType.Int => fields.optionalInt(name)
          case SchemaType.Long => fields.optionalLong(name)
          case SchemaType.String => fields.optionalString(name)
          case SchemaType.Short => fields.optionalInt(name)
          case _ =>
            logger.warn(s"Unknown schema type $schemaType; default to string")
            fields.optionalString(name)
        }
      }.endRecord()
    }

    def configuration: Configuration = {
      val conf = new Configuration
      if (isProjection) {
        AvroReadSupport.setAvroReadSchema(conf, projection)
        AvroReadSupport.setRequestedProjection(conf, projection)
        conf.set(org.apache.parquet.hadoop.ParquetFileReader.PARQUET_READ_PARALLELISM, parallelism)
      }
      conf
    }

    AvroParquetReader.builder[GenericRecord](path)
      .withConf(configuration)
      .withFilter(predicate.fold(FilterCompat.NOOP)(p => FilterCompat.get(p())))
      .build().asInstanceOf[ParquetReader[GenericRecord]]
  }
}

File: eel-core/src/main/kotlin/io/eels/component/parquet/ParquetReaderSupport.kt
package io.eels.component.parquet

import com.typesafe.config.Config
import com.typesafe.config.ConfigFactory
import io.eels.component.Predicate
import io.eels.component.avro.schemaToAvroSchema
import io.eels.util.Logging
import io.eels.util.Option
import io.eels.util.getOrElse
import org.apache.avro.generic.GenericRecord
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
import org.apache.parquet.avro.AvroParquetReader
import org.apache.parquet.avro.AvroReadSupport
import org.apache.parquet.filter2.compat.FilterCompat
import org.apache.parquet.hadoop.ParquetReader

object ParquetReaderSupport : Logging {

  val config: Config = ConfigFactory.load()

  val parallelism = config.getInt("eel.parquet.parallelism").toString().apply {
    logger.debug("Parquet readers will use parallelism = $this")
  }

  /**
   * Creates a new reader for the given path.
   *
   * @predicate if set then a parquet predicate is applied to the rows
   * @projectionSchema if set then the schema is used to narrow the fields returned
   */
  fun create(path: Path,
             predicate: Option<Predicate>,
             projectionSchema: Option<io.eels.schema.Schema>): ParquetReader<GenericRecord> {

    // The parquet reader can use a projection by setting a projected schema onto a conf object
    fun configuration(): Configuration {
      val conf = Configuration()
      projectionSchema.forEach {
        val projection = schemaToAvroSchema(it, false)
        AvroReadSupport.setAvroReadSchema(conf, projection)
        AvroReadSupport.setRequestedProjection(conf, projection)
        conf.set(org.apache.parquet.hadoop.ParquetFileReader.PARQUET_READ_PARALLELISM, parallelism)
      }
      return conf
    }

    // a filter is set when we have a predicate for the read
    fun filter(): FilterCompat.Filter = predicate.map { FilterCompat.get(it.apply()) }.getOrElse { FilterCompat.NOOP }

    @Suppress("UNCHECKED_CAST")
    return AvroParquetReader.builder<GenericRecord>(path)
        .withConf(configuration())
        .withFilter(filter())
        .build() as ParquetReader<GenericRecord>
  }
}